libs/jsonschema/from_type.go

```diff
@@ -0,0 +1,338 @@
+package jsonschema
+
+import (
+	"container/list"
+	"fmt"
+	"maps"
+	"path"
+	"reflect"
+	"slices"
+	"strings"
+)
+
+// Fields tagged "readonly" should not be emitted in the schema as they are
+// computed at runtime, and should not be assigned a value by the bundle author.
+const readonlyTag = "readonly"
+
+// Annotation for internal bundle fields that should not be exposed to customers.
+// Fields can be tagged as "internal" to remove them from the generated schema.
+const internalTag = "internal"
+
+// Annotation for bundle fields that have been deprecated.
+// Fields tagged as "deprecated" are omitted from the generated schema.
+const deprecatedTag = "deprecated"
+
+type constructor struct {
+	// Map of typ.PkgPath() + "." + typ.Name() to the schema for that type.
+	// Example key: github.com/databricks/databricks-sdk-go/service/jobs.JobSettings
+	definitions map[string]Schema
+
+	// Map of typ.PkgPath() + "." + typ.Name() to the corresponding type. Used to
+	// track types that have been seen to avoid infinite recursion.
+	seen map[string]reflect.Type
+
+	// The root type for which the schema is being generated.
+	root reflect.Type
+}
+
+// JSON pointers use "/" as a delimiter to represent nested objects. This means
+// we would instead need to use "~1" to represent "/" if we wish to refer to a
+// key in a JSON object with a "/" in it. Instead of doing that we replace "/" with an
+// additional level of nesting in the output map. Thus the $refs in the generated
+// JSON schema can contain "/" without any issues.
+// see: https://datatracker.ietf.org/doc/html/rfc6901
+//
+// For example:
+// {"a/b/c": "value"} is converted to {"a": {"b": {"c": "value"}}}
+// the $ref for "value" would be "#/$defs/a/b/c" in the generated JSON schema.
+func (c *constructor) Definitions() map[string]any {
+	defs := maps.Clone(c.definitions)
+
+	// Remove the root type from the definitions. We don't need to include it in
+	// the definitions because it will be inlined as the root of the generated JSON schema.
+	delete(defs, typePath(c.root))
+
+	if len(defs) == 0 {
+		return nil
+	}
+
+	res := make(map[string]any)
+	for k, v := range defs {
+		parts := strings.Split(k, "/")
+		cur := res
+		for i, p := range parts {
+			// Set the value for the last part.
+			if i == len(parts)-1 {
+				cur[p] = v
+				break
+			}
+
+			// For all but the last part, create a new map value to add a level
+			// of nesting.
+			if _, ok := cur[p]; !ok {
+				cur[p] = make(map[string]any)
+			}
+			cur = cur[p].(map[string]any)
+		}
+	}
+
+	return res
+}
+
+// FromType converts a reflect.Type to a jsonschema.Schema. Nodes in the final JSON
```

Comment:
Tip: you can wrap type names in `[]` in comments to make them clickable from your IDE.

See https://tip.golang.org/doc/comment#doclinks